File: Roadmap/08 - CLASES/typescript/AChapeton.ts
class Todo {
  constructor(title: string, description: string, done: boolean, user: string, daysLeft: number){
    this.title = title,
    this.description = description,
    this.done = done,
    this.user = user,
    this.daysLeft = daysLeft
  }

  //Los Getters se utilizan para no interactuar directamente con las propiedades de la clase.
  //Tambien se evita para no dar a conocer todas sus propiedades, por seguridad.
  get finishTodo (){
    return this.toggleTodo()
  }

  //Metodos
  toggleTodo() {
    return !this.done
  }
}

const myTodo = new Todo('Finish roadmap', 'Finish exercise #08', false, 'Andres', 5)

console.log(myTodo)

console.log(myTodo.finishTodo)


/* DIFICULTAD EXTRA */

class Pila {
  constructor(data: Array<any>){
    this.data = data
  }

  addElementAction(element: any){
    return this.data.unshift(element)
  }

  deleteElementAction(){
    return this.data.shift()
  }

  sizeAction(){
    return this.data.length
  }

  get addElement() {
    return this.addElementAction
  }

  get deleteElement() {
    return this.deleteElementAction
  }

  get size(){
    return this.sizeAction
  }
}

const nuevaPila = new Pila([])

console.log('Pila vacia', nuevaPila)
nuevaPila.addElement(1)
nuevaPila.addElement(2)
nuevaPila.addElement(3)
console.log('Nuevos elementos agregados', nuevaPila)
console.log('Cantidad de elementos', nuevaPila.size())
nuevaPila.deleteElement()
console.log('Eliminado ultimo elemento', nuevaPila)

class Cola {
  constructor(data: Array<any>){
    this.data = data
  }

  addElementAction(element: any){
    return this.data.push(element)
  }

  deleteElementAction(){
    return this.data.shift()
  }

  sizeAction(){
    return this.data.length
  }

  get addElement() {
    return this.addElementAction
  }

  get deleteElement() {
    return this.deleteElementAction
  }

  get size(){
    return this.sizeAction
  }
}

const nuevaCola = new Cola([])

console.log('Cola vacia', nuevaCola)
nuevaCola.addElement(1)
nuevaCola.addElement(2)
nuevaCola.addElement(3)
console.log('Nuevos elementos agregados', nuevaCola)
console.log('Cantidad de elementos', nuevaCola.size())
nuevaCola.deleteElement()
console.log('Eliminado primer elemento', nuevaCola)

File: Roadmap/08 - CLASES/python/AChapeton.py
class Todo:
  def __init__(self, title, description, done, user, daysLeft):
    self.title = title
    self.description = description
    self.done = done
    self.user = user
    self.daysLeft = daysLeft

  def toggleStatus(self):
    self.done = not self.done
  
  def printTodo(self):
    print(f"""
      Title: {self.title}
      Description: {self.description}
      Done: {self.done}
      User: {self.user}
      Days left: {self.daysLeft}
    """)
  
newTodo = Todo('Test', 'Just testing', False, 'Andres', 4)

newTodo.printTodo()

newTodo.toggleStatus()

newTodo.printTodo()


# DIFICULTAD EXTRA

class Pila:
  def __init__(self, data):
    self.data = data

  def checkData(self):
    print(self.data)

  def addElement(self, value):
    return self.data.append(value)
  
  def deleteElement(self):
    return self.data.pop()

newPila = Pila([1, 2, 3])

newPila.checkData()
newPila.addElement(4)
newPila.addElement(5)
newPila.addElement(6)
newPila.checkData()
newPila.deleteElement()
newPila.checkData()

class Cola:
  def __init__(self, data):
    self.data = data

  def checkData(self):
    print(self.data)

  def addElement(self, value):
    return self.data.append(value)
  
  def deleteElement(self):
    return self.data.pop(0)

newCola = Cola([1, 2, 3])

newCola.checkData()
newCola.addElement(4)
newCola.addElement(5)
newCola.addElement(6)
newCola.checkData()
newCola.deleteElement()
newCola.checkData()